cmd/kyverno/main.go

```diff
@@ -256,22 +264,49 @@ func main() {
 		os.Exit(1)
 	}
 
+	// Metrics Configuration
 	if !disableMetricsExport {
-		promConfig, err = metrics.NewPromConfig(metricsConfigData)
+		if otel == "grpc" {
+			// Otlpgrpc metrics will be served on port 4317: default port for otlpgrpcmetrics
+			metricsAddr := ":" + metricsPort
+			setupLog.Info("Enabling Metrics for Kyverno", "address", metricsAddr)
+
+			endpoint := otelCollector + metricsAddr
+			metricsConfig, err = metrics.NewOTLPGRPCConfig(
+				endpoint,
+				metricsConfigData,
+				transportCreds,
+				kubeClient,
+				log.Log.WithName("Metrics"),
+			)
+
+			if err != nil {
+				setupLog.Error(err, "Failed to enable metrics", "address", ":4317")
+				os.Exit(1)
+			}
+		} else if otel == "prometheus" {
+			// Prometheus Server will serve metrics on metrics-port
+			metricsAddr := ":" + metricsPort
+			setupLog.Info("enabling prometheus metrics", "address", metricsAddr)
+			metricsConfig, metricsServerMux, err = metrics.NewPrometheusConfig(metricsConfigData, log.Log.WithName("Metrics"))
+			go func() {
+				setupLog.Info("Enabling Metrics for Kyverno", "address", metricsAddr)
+				if err := http.ListenAndServe(metricsAddr, metricsServerMux); err != nil {
+					setupLog.Error(err, "failed to enable metrics", "address", metricsAddr)
+					os.Exit(1)
+				}
+			}()
+		}
+	}
+
+	// Tracing Configuration
+	if enableTracing {
+		setupLog.Info("Enabling tracing for Kyverno...")
+		err = tracing.NewTraceConfig(otelCollector, transportCreds, kubeClient, log.Log.WithName("Tracing"))
 		if err != nil {
-			setupLog.Error(err, "failed to setup Prometheus metric configuration")
+			setupLog.Error(err, "Failed to enable tracing for Kyverno")
 			os.Exit(1)
 		}
```

Comment:
Can we extract the whole metrics logic to a separate file? It can be in `cmd/kyverno/metrics.go` or `pkg/metrics/init.go`.

Reply:
Yes we can do that...and then call it into main. We can also wait for when the controllers get refactored...WDYT?

Reply:
What controller refactoring?

Reply:
Sorry I misunderstood this...but yea we can do this.